pandas/io/stata.py

```diff
@@ -1768,14 +1803,39 @@ def _do_convert_categoricals(
         for col, label in zip(data, lbllist):
             if label in value_labels:
                 # Explicit call with ordered=True
-                cat_data = Categorical(data[col], ordered=order_categoricals)
-                categories = []
-                for category in cat_data.categories:
-                    if category in value_label_dict[label]:
-                        categories.append(value_label_dict[label][category])
-                    else:
-                        categories.append(category)  # Partially labeled
+                vl = value_label_dict[label]
+                keys = np.array(list(vl.keys()))
+                column = data[col]
+                key_matches = column.isin(keys)
+                if self._chunksize is not None and key_matches.all():
+                    initial_categories = keys
+                    # If all categories are in the keys and we are iterating,
+                    # use the same keys for all chunks. If some are missing
+                    # value labels, then we will fall back to the categories
+                    # varying across chunks.
+                else:
+                    if self._chunksize is not None:
```

Comment:
you *could* do a pass over the file to collect the categories then re-read them (is this relatively cheap to do?)

Reply:
It could be pretty expensive if the file as 10M rows and many columns.  

One *could* write a function that only gets the values of specific variables without reading the entire file contents, which would be faster than full passes. But more work.  Maybe leave this open and I'll see how hard this special case it.

Reply:
kk, lgtm otherwise, (just rebase to fix the conflict), ping when ready.

Reply:
I've thought through this issue and I've concluded that it isn't "safe" to try and read all categories first.  Worse case is that a column has random values so that all values are distinct. This would require storing the entire column. If a single column exceeds the memory of the machine being used, then it will crash.

This is a pretty extreme scenario, but can't be ruled out. It seems best to warn users that they might need to "glue" their categoricals together themselves.  I expect in most cases that users will have fully labeled columns and so will not see this warning.